frontend/test/pytest/test_callback.py

```diff
@@ -811,6 +812,330 @@ def wrapper_jit(x):
+    x = jnp.array([[1.0, 0.5], [0.12, -1.2]])
+    y = jnp.array([-0.6, 0.2])
+    flat_results_obs, _ = jax._src.tree_util.tree_flatten(result(x, y))
+    flat_results_exp, _ = jax._src.tree_util.tree_flatten(expected(x, y))
+    for obs, exp in zip(flat_results_obs, flat_results_exp):
+        assert np.allclose(obs, exp)
+
+    # (Array([[2.5208345e-02, 4.5493888e-03],
+    #         [5.7185785e-04, 2.3190898e-05]], dtype=float64),
+    #  Array([-0.40939555,  0.02444299], dtype=float64))
+
+
+def test_pytree_in_pytree_out():
+    """Test pytree in pytree out"""
+
+    shape = {
+        "one": jax.ShapeDtypeStruct((2,), jnp.float64),
+        "two": jax.ShapeDtypeStruct((2,), jnp.float64),
+    }
+
+    @pure_callback
+    def some_func(weights) -> shape:
+        return {"one": np.sin(weights["x"]) @ weights["y"], "two": weights["y"] ** 2}
+
+    @some_func.fwd
+    def some_func_fwd(weights):
+        return some_func(weights), weights
+
+    @some_func.bwd
+    def some_func_bwd(res, dy):
+        vjp0 = res["y"] * jnp.cos(res["x"]) * jnp.reshape(dy["one"], (-1, 1))
+        vjp1 = dy["one"] @ jnp.sin(res["x"]) + 2 * res["y"] * dy["two"]
+        return ({"x": vjp0, "y": vjp1},)
+
+    @qml.qjit
+    @grad
+    def result(weights):
+        weights["y"] = weights["y"] ** 2
+        res = some_func(weights)
+        return jnp.dot(res["one"], res["two"])
+
+    @jax.jit
+    @jax.grad
+    def expected(weights):
+        x = weights["x"]
+        y = weights["y"]
+        return jnp.dot(jnp.sin(x) @ y**2, y**4)
+
+    weights = {"x": jnp.array([[1.0, 0.5], [0.12, -1.2]]), "y": jnp.array([-0.6, 0.2])}
+    flat_results_obs, _ = jax._src.tree_util.tree_flatten(result(weights))
+    flat_results_exp, _ = jax._src.tree_util.tree_flatten(expected(weights))
+    for obs, exp in zip(flat_results_obs, flat_results_exp):
+        assert np.allclose(obs, exp)
+
+    # {'x': Array([[2.5208345e-02, 4.5493888e-03],
+    #        [5.7185785e-04, 2.3190898e-05]], dtype=float64),
+    # 'y': Array([-0.40939555,  0.02444299], dtype=float64)}
+
+
+def test_callback_backwards_function():
+    """Test a workflow where the bwd function is also using
+    a callback to compute the VJP"""
+
+    shape = {
+        "one": jax.ShapeDtypeStruct((2,), jnp.float64),
+        "two": jax.ShapeDtypeStruct((2,), jnp.float64),
+    }
+    vjp_shape = {
+        "x": jax.ShapeDtypeStruct((2, 2), jnp.float64),
+        "y": jax.ShapeDtypeStruct((2,), jnp.float64),
+    }
+
+    @pure_callback
+    def some_func(weights) -> shape:
+        return {"one": np.sin(weights["x"]) @ weights["y"], "two": weights["y"] ** 2}
+
+    @some_func.fwd
+    def some_func_fwd(weights):
+        return some_func(weights), weights
+
+    @pure_callback
+    def some_func_bwd_vjp(res, dy) -> vjp_shape:
+        vjp0 = res["y"] * jnp.cos(res["x"]) * jnp.reshape(dy["one"], (-1, 1))
+        vjp1 = dy["one"] @ jnp.sin(res["x"]) + 2 * res["y"] * dy["two"]
+        return ({"x": vjp0, "y": vjp1},)
```

Comment:
@erick-xanadu since this is a pure callback, I thought it would be important to use `np.` here (since we know numpy won't work without the purecallback)